R/utils.R

```diff
@@ -165,3 +169,123 @@ random_name <- function(prefix = "") {
   name <- paste0(sample(vals, 10, replace = TRUE), collapse = "")
   paste0(prefix, "odbc_", name)
 }
+
+# apple + spark drive config (#651) --------------------------------------------
+configure_spark <- function() {
+  if (is_windows()) {
+    return(invisible())
+  }
+
+  if (!has_unixodbc()) {
+    install_unixodbc()
+  }
+
+  unixodbc_install <- locate_install_unixodbc()
+  if (identical(unixodbc_install, character(0))) {
+    unixodbc_install <- install_unixodbc()
+  }
+
+  spark_config <- locate_config_spark()
+  if (identical(spark_config, character(0))) {
+    abort(c(
+      "Unable to locate the needed spark ODBC driver.",
+      i = "Please install the needed driver from https://www.databricks.com/spark/odbc-drivers-download."
+    ))
+  }
+
+  configure_unixodbc_spark(unixodbc_install[1], spark_config[1])
+}
+
+locate_install_unixodbc <- function() {
+  unixodbc_prefix <- tryCatch(
+    system("odbc_config --lib-prefix", intern = TRUE),
+    error = function(e) character(0)
+  )
+  if (!identical(unixodbc_prefix, character(0))) {
+    return(paste0(unixodbc_prefix, "/libodbcinst.dylib"))
+  }
+
+  common_dirs <- c(
+    "/usr/local/lib",
+    "/opt/homebrew/lib",
+    "/opt/homebrew/opt/unixodbc/lib"
+  )
+
+  list.files(
+    common_dirs,
+    pattern = "libodbcinst\\.dylib$",
+    full.names = TRUE
+  )
+}
+
+install_unixodbc <- function(call = caller_env()) {
+  tryCatch(
+    {install_unixodbc_libs()},
+    error = function(e) {
+      abort(
+        c(
+          "Unable to install the unixODBC driver manager.",
+          i = "Please install unixODBC using Homebrew with `brew install unixodbc`."
+        ),
+        call = call
+      )
+    }
+  )
+
+  locate_install_unixodbc()
+}
+
+install_unixodbc_libs <- function() {
+  source("https://mac.R-project.org/bin/install.R")
+  install.libs("unixodbc")
```

Comment:
I personally wasn't able to successfully run these lines on either my work computer (M1, existing unixodbc install) or personal computer (Intel, no existing unixodbc install). Both failed with errors:

```
various/file/paths: Can't restore time
```

I _think_ this approach would also require some ODBCSYSINI magic, as well?

Reply:
I think it's fine to just tell the user what they need. 

But maybe checking for `unixODBC` should happen `onLoad()` since this will affect all connection?

Reply:
Heard. Just transitioned from trying to install unixODBC to erroring if not installed in 60cc4da5.

Agreed that checking for unixODBC on load is reasonable. Should that happen in this PR or a separate one?